Documentation/design-docs/standalone-gc-eventing.md

```diff
@@ -0,0 +1,212 @@
+performance events by a standalone GC. This system must have two properties in order to be
+acceptable:
+
+1. It must be *efficient* to query whether or not a particular event is turned on. It is not acceptable
+to perform an indirection (i.e. cross the GC/EE interface boundary) in order to get this information.
+2. The cost of firing an event by a standalone GC should be comparable to the cost of firing an event
+without using a standalone GC.
+
+## Querying Whether Events Are Enabled
+
+Tt is not acceptable to perform an indirection when querying whether or not a requested event
+is enabled. Therefore, it follows that the GC must maintain some state about what events are currently 
+enabled. Events are enabled through *keywords* and *levels*; a particular event is enabled if both the 
+event's keyword and the event's level have been enabled on the provider to which the event belongs.
+
+To accomplish this, the GC will contain a class with this signature:
+
+```c++
+// Constructed on GC startup for every event that the GC can fire.
+struct GCEvent
+{
+    /* opaque, but will contain information about what event this is,
+       its keyword, its level, and a handle to the EE's representation of this event */
+    /* this structure is potentially auto-generated from an event manifest */
+};
+
+class GCEventStatus
+{
+public:
+    // Returns true if this event is currently enabled, false otherwise.
+    static bool IsEventEnabled(const GCEvent& event);
+
+    // Enables events with the given keyword and level.
+    static void Enable(int keyword, int level);
+
+    // Disables events the given keyword and level.
+    static void Disable(int keyword, int level);
+};
+
+```
+
+The GC will use `GCEventStatus::IsEventEnabled` to query whether or not a particular event is enabled.
+Whenever the EE observes a change in what keywords or levels are enabled, it must inform the GC of the
+change so that it can update `GCEventStatus` using `Enable` and `Disable`. The exact mechanism by which
+the EE observes a change in the event state is described further below. ("Getting Informed of Changes to 
+Event State")
+
+When the EE *does* observe a change, it informs the GC using these two new methods on `IGCHeap`:
+
+```c++
+void IGCHeap::EnableEvents(int keyword, int level)
+void IGCHeap::DisableEvents(int keyword, int level)
+```
+
+
+These interface methods in turn will call `GCEventStatus::{Enable, Disable}` in order
+to enable or disable keywords or levels. The currently enabled keywords and levels are encoded as
+bit vectors so that querying whether an event is enabled is as fast as loading three globals, a pointer
+dereference, and two bit tests, e.g. (hypothetically):
+
+```c++
+uint32_t enabledLevels;
+uint32_t enabledKeywords;
+
+bool GCEventStatus::IsEventEnabled(const GCEvent& event)
+{
+    return (enabledLevels & event.level) && (enabledKeywords & event.keyword);
```

Comment:
You can also do what the EventPipe does, which is keep a bit per event (as part of GCEvent) and when a change is made to the set of keywords, you re-calculate the value of the bit for each GCEvent.  Then you can implement IsEnabled as:

```
bool GCEventStatus::IsEnabledEvent(const GCEvent& event)
{
    return event.m_isEnabled;
}
```